### ISSUE TITLE
functorch.functionalize doesn't error out with logcumsumexp.out

### ISSUE THREAD

**Issue by alanwaketan** [pytorch/pytorch#92912]
### 🐛 Describe the bug

logcumsumexp.out will error out if the given output buffer's dtype is different from the input, but somehow that property is violated after functionalization, here is the PoC:
```
import torch
import functorch


def poc6():
  device = 'cpu'
  axis = -1

  # Check input and inplace_output type mismatch
  b = torch.randn(5, 2, device=device, dtype=torch.float64)
  inplace_out = torch.zeros(5, 2, device=device, dtype=torch.float32)
  try:
    torch.logcumsumexp(b, axis, out=inplace_out)
  except Exception as e:
    print(e)



print("CPU")
poc6()
print()

print("CPU Functionalize")
functorch.functionalize(poc6)()
```

### Versions

Nightly.

**Comment by bdhirsh** [comment]
Hmm. So what's going on here is that `logcumsumexp.out` has some extra shape/dtype checks, that its functional variant does not have. Those checks live [here](https://github.com/pytorch/pytorch/blob/acdd462b1a070790799ce4623ce8ecc83e197e81/aten/src/ATen/native/ReduceOps.cpp#L474).

One way to get those checks to run properly would be to run the out= op with meta tensors, before redispatching to the functional op. In fact, we already do that for inplace ops, but we don't do it for out= ops.

There's a comment explaining the situation here: https://github.com/pytorch/pytorch/blob/acdd462b1a070790799ce4623ce8ecc83e197e81/torchgen/gen_functionalization_type.py#L614

What makes it more annoying to do this properly for out= ops is that not all out= ops have a meta function. Even worse, we don't have a way of statically knowing which ops do/don't have meta functions, because some ops only have a composite implementation (and we'd need to know if their decomposed ops all have meta functions).

That was the original reason that we didn't end up adding those checks. One thing I'm not sure about though, is that it's possible our meta coverage has gotten a lot better since then. It could be worth adding that check, and running the suite to see if there are any failures.

One question I have for XLA - how critical do you think it is to get these checks back @alanwaketan?

**Comment by alanwaketan** [comment]
@bdhirsh No, not critical at all. I can mark it not to be fixed for now. Just trying to report my findings. 😄

**Comment by ysiraichi** [comment]
#### (1)

What if we had a tool that generated a list of operations that implement `out=` meta functions? Then, when actually generating the functional kernels, we could allow meta function execution for them. In summary:

- Create a `tools/get_ops_with_out_meta_impl.py` that runs each `out=` operation in `native_functions.yaml` on the meta device, returning a list of operations that succeeded. This list would be updated (i.e. the tool would be run) only when a new meta function is implemented (due to composite ops). Therefore, it's overhead isn't that important.
- Allow the meta function corresponding to the `out=` operations on that list to run at the beginning of the functionalization kernel (as is, today).

-----

#### (2)

The solution above should take care of most operations that present this error. For those `out=` operations that don't have a meta function, we could try to come up with some heuristic for deciding whether [a "safe-copy"](https://github.com/pytorch/pytorch/wiki/Developer-FAQ#how-does-out-work-in-pytorch) or a dtype check would be needed. This heuristic could work with `force_safe_copy_ops` and `force_dtype_check_ops` lists. For example, the heuristic could be:

> All ops that take only a single tensor (no scalar) argument should have a dtype check.

**Reasoning:** they definitely didn't go through dtype-promotion. However, there are some operations, such as `index.Tensor` that takes in 2 tensor arguments, but doesn't go through dtype-promotion. So, that would be one operation to be in the `force_dtype_check_ops` list.

-----

That said, I think we should implement (1) first, and see what kind and how many operations would fall under (2). Depending on the number, we could even do it manually, without a heuristic.

@bdhirsh Let me know what you think.

**Comment by ysiraichi** [comment]
I see. That makes sense. That means that maybe #138399 is of interest, here.
